cocotb/share/lib/embed/gpi_embed.c

```diff
@@ -230,214 +240,115 @@ int embed_sim_init(gpi_sim_info_t *info)
         return ret;
 
-    // Find the simulation root
-    const char *dut = getenv("TOPLEVEL");
-
-    if (dut != NULL) {
-        if (!strcmp("", dut)) {
-            /* Empty string passed in, treat as NULL */
-            dut = NULL;
-        } else {
-            // Skip any library component of the toplevel
-            char *dot = strchr(dut, '.');
-            if (dot != NULL) {
-                dut += (dot - dut + 1);
-            }
-        }
-    }
-
-    PyObject *cocotb_module, *cocotb_init, *cocotb_retval;
-    PyObject *cocotb_log_module = NULL;
+    PyObject *cocotb_module = NULL;
+    PyObject *cocotb_gpi_module = NULL;
     PyObject *simlog_func;
-    PyObject *argv_list;
-
-    cocotb_module = NULL;
+    PyObject *entry_tuple = NULL;
+    PyObject *cocotb_init = NULL;
+    PyObject *argv_list = NULL;
+    PyObject *cocotb_retval = NULL;
 
     // Ensure that the current thread is ready to call the Python C API
     PyGILState_STATE gstate = PyGILState_Ensure();
     to_python();
 
-    if (get_module_ref("cocotb", &cocotb_module))
-        goto cleanup;
-
-    if (get_module_ref("cocotb.log", &cocotb_log_module)) {
+    if (get_module_ref("cocotb._gpi_embed", &cocotb_gpi_module)) {
         goto cleanup;
     }
 
     // Obtain the function to use when logging from C code
-    simlog_func = PyObject_GetAttrString(cocotb_log_module, "_log_from_c");      // New reference
+    simlog_func = PyObject_GetAttrString(cocotb_gpi_module, "_log_from_c");      // New reference
     if (simlog_func == NULL) {
         PyErr_Print();
         LOG_ERROR("Failed to get the _log_from_c function");
         goto cleanup;
     }
-    if (!PyCallable_Check(simlog_func)) {
-        LOG_ERROR("_log_from_c is not callable");
-        Py_DECREF(simlog_func);
-        goto cleanup;
-    }
 
-    set_log_handler(simlog_func);                                       // Note: This function steals a reference to simlog_func.
+    pLogHandler = simlog_func;                                          // Note: This function steals a reference to simlog_func.
 
     // Obtain the function to check whether to call log function
-    simlog_func = PyObject_GetAttrString(cocotb_log_module, "_filter_from_c");   // New reference
+    simlog_func = PyObject_GetAttrString(cocotb_gpi_module, "_filter_from_c");   // New reference
     if (simlog_func == NULL) {
         PyErr_Print();
         LOG_ERROR("Failed to get the _filter_from_c method");
         goto cleanup;
     }
-    if (!PyCallable_Check(simlog_func)) {
-        LOG_ERROR("_filter_from_c is not callable");
-        Py_DECREF(simlog_func);
+
+    pLogFilter = simlog_func;                                           // Note: This function steals a reference to simlog_func.
+
+    gpi_set_log_handler(cocotb_log_v, NULL);
+
+    entry_tuple = PyObject_CallMethod(cocotb_gpi_module, "_load_entry", NULL);
+    if (entry_tuple == NULL) {
+        PyErr_Print();
+        LOG_ERROR("Unable to load entry point");
```

Comment:
Printing the entry point name here could aid debugging.

Reply:
I added a try-catch block in Python to add the COCOTB_ENTRY string to the exception that gets reported here.